http_handlers.go

```diff
@@ -102,6 +141,109 @@ func getMetadata(filter string) http.HandlerFunc {
 	}
 }
 
+func ec2Handler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	logger.Debug("Calling ec2Handler ")
+	userIP := getIPFromRequest(r)
+	if userIP == "" {
+		return
+	}
+
+	metrics.MetadataRequests.Inc()
+	logger.With("userIP", userIP).Info("Actual IP is: ")
+	hw, err := getByIP(context.Background(), hegelServer, userIP) // returns hardware data as []byte
+	if err != nil {
+		metrics.Errors.WithLabelValues("metadata", "lookup").Inc()
+		logger.Info("Error in finding hardware: ", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	res, err := processEC2Query(r.URL.Path)
+	if err != nil {
+		logger.Error(err)
+		w.WriteHeader(http.StatusNotFound)
+		_, err := w.Write([]byte("404 not found"))
+		if err != nil {
+			logger.Error(err, "failed to write error response")
+		}
+	}
+
+	var resp []byte
+	if filter, ok := res.(string); ok {
+		resp, err = filterMetadata(hw, filter)
+		if err != nil {
+			logger.Error(err, "Error in filtering metadata: ")
+		}
+	} else if submenu, ok := res.(map[string]interface{}); ok {
+		var keys []string
+		for item := range submenu {
+			switch item {
+			case "_base": // _base is only used to keep track of the base filter, not a metadata item
+				continue
+			case "spot": // list only if instance is spot
+				resp, err := filterMetadata(hw, ".metadata.instance.spot")
+				if err != nil {
+					logger.Info("Error in filtering metadata: ", err)
+				}
+				if string(resp) != "" {
+					keys = append(keys, item)
+				}
+			default:
+				keys = append(keys, item)
+			}
+		}
+
+		sort.Strings(keys)
+
+		for _, item := range keys {
+			resp = []byte(fmt.Sprintln(string(resp) + item))
+		}
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(resp)
+	if err != nil {
+		logger.Error(err, "failed to write Metadata")
+	}
+}
+
+// processEC2Query returns either a specific filter (used to parse hardware data for the value of a specific field),
+// or a map of filters (used for printing out its keys)
+func processEC2Query(query string) (interface{}, error) {
```

Comment:
does it make sense to do a unit test around this function?

Reply:
Sure, I can add one in

Reply:
Please consider handling all these use cases as well
```
if filter, ok := item.(string); ok { // if is an actual filter
			result = fmt.Sprint(base, filter)
		} else if subfilters, ok := item.(map[string]interface{}); ok { // if is another map of filters
			base = fmt.Sprint(base, subfilters["_base"].(string))
			result = subfilters
		} else {
			return nil, errors.New("invalid metadata item")
```

Reply:
Hmm is there a specific case you're thinking of? I can't think of anything I'm missing here https://github.com/tinkerbell/hegel/blob/2e1275844da2e97db4c4513bf3219f83b2458db2/http_handlers_test.go#L513

The first two test cases test when `item` is a string, the third one tests if `item`, the last three tests the error cases.